Making templates more safe

Qubad786 opened the pull request:
@mushtaqak , @robrap may you guys please review this? 

Thanks!

mushtaqak commented on lms/templates/courseware/course_about.html:
```diff
@@ -134,7 +136,10 @@
         %elif in_cart:
           <span class="add-to-cart">
-            ${_('This course is in your <a href="{cart_link}">cart</a>.').format(cart_link=cart_link)}
+            ${_('This course is in your {link_start}cart{link_end}.').format(
```
Here you need Text and HTML methods too

robrap commented:
This is a general message being posted for all Safe Templates PRs.  If you are making steady progress, sorry about this, and thank you!

We need to get any Safe Template PRs merged as soon as possible.

Make sure you have the latest copy of `./scripts/safe_template_linter.py` and `./scripts/safe-commit-linter.sh`, and make sure that `./scripts/safe-commit-linter.sh` reports 0 violations for this PR.

Feel free to ping me with any questions, and/or refer to the [Safe Template Linter documentation](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter ).

Thanks!

Qubad786 commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
-        $('#register_error').html(
-            (xhr.responseText ? xhr.responseText : "${_("An error occurred. Please try again later.")}")
+        $('#register_error').text(
+            (xhr.responseText ? xhr.responseText : "${_('An error occurred. Please try again later.')| n, js_escaped_string}")
```
I never wanted to use `js_escaped_string` but ended up doing this to make linter happy. @robrap should we escape here?

robrap replied:
Assuming `xhr.responseText` is plain text (i.e. doesn't contain HTML tags), then this is perfect.  The browser does multiple passes over this code.  The first pass it is trying to determine what is HTML and what is script.  The `js_escaped_string` makes sure that everything looks like JavaScript and doesn't mistakenly look like HTML.  The call to `text()` is used by JavaScript to add text to the DOM, and again if anything looks like HTML, it will take care of making sure it renders as text.

Read this brief section in the docs for more understanding about [escaping in JavaScript contexts](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#javascript-context-and-escaping).

Nit: I'd surround ` | ` with spaces on both sides.  

Also, you have two comments here, but I just see the one line.

Qubad786 replied:
Thanks! 2nd comment was for very similar context – L105.

mushtaqak replied:
nit: space before and after `|`

Qubad786 commented on lms/templates/courseware/course_about.html:
```diff
@@ -83,16 +93,16 @@
-        $('#register_error').html(
-            (xhr.responseText ? xhr.responseText : "${_("An error occurred. Please try again later.")}")
+        $('#register_error').text(
+            (xhr.responseText ? xhr.responseText : "${_('An error occurred. Please try again later.')| n, js_escaped_string}")
```
Again. @robrap should we escape here?

Qubad786 commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
-        $('#register_error').html("${perms_error}").css("display", "block");
+          location.href = "${reverse('course-specific-register', args=[course.id.to_deprecated_string()])| n, js_escaped_string}?course_id=${course.id| n, js_escaped_string}&enrollment_action=enroll";
+      } else if (xhr.status == 400) {//This means the user did not have permission
+          edx.HtmlUtils.setHtml(
```
@robrap , I don't know of any other way of using `HtmlUtils` in mako templates. I am wondering if that's the right way?

robrap replied:
This is the correct way to use it where RequireJS is not available, so this is perfect.  Thanks.

robrap commented:
Thanks for this work.  I'll respond to your comments, but wanted to note the linter has one more violation which you can see in the [Build Artifacts report](https://build.testeng.edx.org/job/edx-platform-quality-pr/18591/artifact/edx-platform/reports/safecommit/safecommit.report/*view*/).

robrap commented on cms/templates/import.html:
```diff
@@ -10,6 +11,7 @@
 <%namespace name='static' file='static_content.html'/>
 <%!
   from django.utils.translation import ugettext as _
+  from openedx.core.djangolib.markup import Text, HTML
```
Nit: Order should be "HTML, Text".

robrap commented on cms/templates/import.html:
```diff
@@ -44,11 +46,11 @@ <h1 class="page-header">
         ## Translators: ".tar.gz" is a file extension, and files with that extension are called "gzipped tar files": these terms should not be translated
         %if library:
-            <p>${_("Be sure you want to import a library before continuing. The contents of the imported library will replace the contents of the existing library. {em_start}You cannot undo a library import{em_end}. Before you proceed, we recommend that you export the current library, so that you have a backup copy of it.").format(em_start='<strong>', em_end="</strong>")}</p>
+            <p>${Text(_("Be sure you want to import a library before continuing. The contents of the imported library will replace the contents of the existing library. {em_start}You cannot undo a library import{em_end}. Before you proceed, we recommend that you export the current library, so that you have a backup copy of it.")).format(em_start=HTML('<strong>'), em_end=HTML('</strong>'))}</p>
```
Nit: I prefer seeing the format parameters, at a minimum, on a next line.

Qubad786 commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
-                          start_help_tag="<a href='{url}'>".format(url=marketing_link('FAQ')), end_tag='</a>',
-                          start_logout_tag="<a href='{url}'>".format(url=reverse('logout'))
-                          )
+        perms_error = Text(
```
Referring [Builder Artifact Report](https://build.testeng.edx.org/job/edx-platform-quality-pr/18591/artifact/edx-platform/reports/safecommit/safecommit.report/*view*/), @robrap I know about this but that's weird as we've made `perms_error` safe here, we cannot afford to escape this again when we use it – on L82.

Qubad786 replied:
@robrap the above comment is about that 1 violation.

robrap commented on lms/templates/courseware/course_about.html:
```diff
@@ -83,16 +93,16 @@
       if(xhr.status == 200) {
         if (xhr.responseText == "") {
-          location.href = "${reverse('dashboard')}";
+          location.href = "${reverse('dashboard')| n, js_escaped_string}";
```
Nit: Add spaces before the `|` in the lines below.

robrap commented:
@Qubad786: Looks good.  A couple of nits, and one missed violation.  Be sure to check your work manually, which it looks like you did to know to add `HTML()` in certain locations.  Thanks!

robrap commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
+      } else if (xhr.status == 400) {//This means the user did not have permission
+          edx.HtmlUtils.setHtml(
+                  $('#register_error'),
+                  "${perms_error}"
```
In new Mako code, we would only have the minimal JavaScript to pass data to actual JavaScript files and Underscore.js templates.  In this legacy code, mixing HTML, JavaScript, and Mako template syntax can get confusing.

Imagine that this code were actually:
```
edx.HtmlUtils.setHtml(
    $('#register_error'),
    "<strong>Message with HTML.</strong>"
)
``` 
Then it may be more clear that `HtmlUtils.HTML()` is needed to communicate that this is already HTML to the `HtmlUtils.setHtml()` call.
```
edx.HtmlUtils.setHtml(
    $('#register_error'),
    edx.HtmlUtils.HTML("<strong>Message with HTML.</strong>")
)
``` 

Now, if we insert back the Mako expression, we get:
```
edx.HtmlUtils.setHtml(
    $('#register_error'),
    edx.HtmlUtils.HTML("${perms_error | n, js_escaped_string}")
)
``` 
Let me know if this works for you.  Note that js_escaped_string is not performing HTML escaping.  Instead, it is performing JavaScript escaping.  It can be confusing at first, but you'll get more and more used to it.

Thanks for making the issue more clear.

Qubad786 replied:
Thanks for clarification, I'll follow that!

robrap replied:
@Qubad786: Hopefully this makes it even more clear the difference between JavaScript escaping and HTML escaping.  Let me know if this helps or is just confusing.

Python code showing JavaScript escaping:
```
>>> from openedx.core.djangolib.js_utils import js_escaped_string
>>> print js_escaped_string("<strong>Message with HTML.</strong>")
\u003Cstrong\u003EMessage with HTML.\u003C/strong\u003E
```

JQuery code showing how a JavaScript-escaped string is not HTML-escaped:
```
$('body').append("\u003Cstrong\u003EMessage with HTML.\u003C/strong\u003E")

<body>
...
<strong>Message with HTML.</strong>
</body>
```

Qubad786 replied:
@robrap that makes perfect sense. Thanks!

Qubad786 commented:
Thanks @robrap. I've been checking manually too, I will be bugging you again after addressing your comments :)

@mushtaqak may you please review as well. Thanks!

mushtaqak commented on cms/templates/import.html:
```diff
@@ -44,11 +46,11 @@ <h1 class="page-header">
             <p>${_("The import process has five stages. During the first two stages, you must stay on this page. You can leave this page after the Unpacking stage has completed. We recommend, however, that you don't make important changes to your library until the import operation has completed.")}</p>
         %else:
-            <p>${_("Be sure you want to import a course before continuing. The contents of the imported course will replace the contents of the existing course. {em_start}You cannot undo a course import{em_end}. Before you proceed, we recommend that you export the current course, so that you have a backup copy of it.").format(em_start='<strong>', em_end="</strong>")}</p>
+            <p>${Text(_("Be sure you want to import a course before continuing. The contents of the imported course will replace the contents of the existing course. {em_start}You cannot undo a course import{em_end}. Before you proceed, we recommend that you export the current course, so that you have a backup copy of it.")).format(em_start=HTML('<strong>'), em_end=HTML('</strong>'))}</p>
```
format placeholders on new line

mushtaqak commented on cms/templates/register.html:
```diff
@@ -1,7 +1,9 @@
 <%def name="online_help_token()"><% return "register" %></%def>
 <%!
 from django.utils.translation import ugettext as _
+from openedx.core.djangolib.markup import Text, HTML
```
nit: import oder ; HTML before Text import

mushtaqak commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
     $('#class_enroll_form').on('ajax:complete', function(event, xhr) {
       if(xhr.status == 200) {
-        location.href = "${reverse('dashboard')}";
+          location.href = "${reverse('dashboard')| n, js_escaped_string}";
```
remove extra space

mushtaqak commented on lms/templates/problem_ajax.html:
```diff
@@ -1 +1,2 @@
-<div id="problem_${element_id}" class="problems-wrapper" data-problem-id="${id}" data-url="${ajax_url}" data-progress_status="${progress_status}" data-progress_detail="${progress_detail}" data-content="${content | h}"></div>
+<%page expression_filter="h"/>
+<div id="problem_${element_id}" class="problems-wrapper" data-problem-id="${id}" data-url="${ajax_url}" data-progress_status="${progress_status}" data-progress_detail="${progress_detail}" data-content="${content}"></div>
```
line return here.

mushtaqak commented:
Looking good, some nits than this would be ready :)

robrap commented on lms/templates/courseware/course_about.html:
```diff
@@ -31,49 +36,54 @@
       add_course_complete_handler = function(jqXHR, textStatus) {
         if (jqXHR.status == 200) {
-          location.href = "${cart_link}";
+          location.href = "${cart_link| n, js_escaped_string}";
```
Nit: Note that all of these should have a space before the `|`.  It would be great if you could search for and fix all of those before the next pass.  Thanks.

Qubad786 commented:
@robrap , after removing that violation, 2 tests are failing, `test_anonymous_user_shib_course` and `test_logged_in_shib_course`. In these tests, `perms_error` is expected to be in response content but response content contains the following with unicode values: 
```
edx.HtmlUtils.setHtml(\n                  $(\'#register_error\'),\n                  edx.HtmlUtils.HTML("The currently logged\\u002Din user account does not have permission to enroll in this course. You may need to \\u003Ca href\\u003D\\u0027/logout\\u0027\\u003Elog out\\u003C/a\\u003E then try the enroll button again. Please visit the \\u003Ca href\\u003D\\u0027#\\u0027\\u003Ehelp page\\u003C/a\\u003E for a possible solution.")\n          )\n
```
What do we do with this? Thanks!

robrap commented:
Assuming this all works correctly, which I think it should, I think the test just needs to be updated.  You may want to call js_escaped_string on the string before it is compared.  Or, you could hard-code the escaped character into the string you are looking for.

I'm not sure if this will break other tests or not.  You'll see.

Qubad786 commented:
@robrap , @mushtaqak I've addressed all the comments.

robrap commented on lms/djangoapps/courseware/tests/test_about.py:
```diff
@@ -39,7 +39,7 @@
 # HTML for registration button
 REG_STR = "<form id=\"class_enroll_form\" method=\"post\" data-remote=\"true\" action=\"/change_enrollment\">"
-SHIB_ERROR_STR = "The currently logged-in user account does not have permission to enroll in this course."
+SHIB_ERROR_STR = "The currently logged\\u002Din user account does not have permission to enroll in this course."
```
Nit: The following is probably more clear without requiring explanation:
```
SHIB_ERROR_STR = js_escaped_string("The currently logged-in user account does not have permission to enroll in this course.")
```
Also, was this error tested manually?  Just curious.

robrap commented:
👍  I made one minor comment that would make the test more clear.  Feel free to make the change when you squash.  I don't need to re-review.  Thanks!

mushtaqak commented:
👍

robrap commented:
@Qubad786 @mushtaqak: We need to merge or close these old "safe templates" PRs.

This one would be a shame to lose since it had two thumbs and should have just been merged.  If you can get this merged but there will be some delay, please add a supporting JIRA ticket and add it to the PR.

Thank you!

mushtaqak commented:
@Qubad786 Closing this PR, if you are working on it, you can open a new PR :)
